utils/io-wrappers.cc

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright (C) 2025-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
+ */
+
+#include "io-wrappers.hh"
+#include <seastar/util/internal/iovec_utils.hh>
+
+using namespace seastar;
+
+class noop_file_impl : public file_impl {
+public:
+    [[noreturn]] void not_implemented() const {
+        throw std::logic_error("unsupported operation");
+    }
+    future<size_t> write_dma(uint64_t pos, const void* buffer, size_t len, io_intent*) override {
+        not_implemented();
+    }
+    future<size_t> write_dma(uint64_t pos, std::vector<iovec> iov, io_intent* intent) override {
+        not_implemented();
+    }
+    future<size_t> read_dma(uint64_t pos, void* buffer, size_t len, io_intent*) override {
+        not_implemented();
+    }
+    future<size_t> read_dma(uint64_t pos, std::vector<iovec> iov, io_intent*) override {
+        not_implemented();
+    }
+    future<> flush(void) override {
+        return make_ready_future<>();
+    }
+    future<struct stat> stat(void) override {
+        not_implemented();
+    }
+    future<> truncate(uint64_t length) override {
+        not_implemented();
+    }
+    future<> discard(uint64_t offset, uint64_t) override {
+        not_implemented();
+    }
+    future<> allocate(uint64_t position, uint64_t) override {
+        not_implemented();
+    }
+    future<uint64_t> size(void) override {
+        not_implemented();
+    }
+    future<> close() override {
+        return make_ready_future<>();
+    }
+    std::unique_ptr<seastar::file_handle_impl> dup() override {
+        not_implemented();
+    }
+    subscription<directory_entry> list_directory(std::function<future<> (directory_entry de)> next) override {
+        not_implemented();
+    }
+    future<temporary_buffer<uint8_t>> dma_read_bulk(uint64_t offset, size_t range_size, io_intent* intent) override {
+        not_implemented();
+    }
+};
+
+file create_file_for_sink(data_sink sink) {
```

Comment:
AFAIU this is used by `file_io_extension::wrap_sink` and `encryption_file_io_extension::wrap_sink` to effectively wrap one sink into another, and this file is used as a transparent connector between those two sinks. Why not make the outer sink wrap the inner sink directly, without this file in the middle?

Reply:
Because we assume an existing implementation has only create a file_impl-based transform. A sink->wrapper-file->file-sink-wrapper->sink chain again ensures existing code continues to work. Less than ideally efficient, but work.

Reply:
Changes later ensures we do the wrapping efficiently where it matters.